# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: js/src/debugger/Debugger.cpp
# Commit: 2d470b425ce9
# Full Hash: 2d470b425ce9d8fc12c89f087d76ff1a337619ab
# Author: Logan Smyth <loganfsmyth@gmail.com>
# Date: 2020-06-15 21:48:38
# Description:
#   Bug 1644699 - Allow Debugger.Frame to have GeneratorInfo that is closed. r=arai
#   
#   The GeneratorInfo data associated with a Debugger.Frame is not guaranteed
#   to be cleared when a generator transitions from running/suspended to closed.
#   In the current codebase, we have broadly assumed that to be true, which was
# ==============================================================================

diff -r d56e40dc9c0f -r 2d470b425ce9 js/src/debugger/Debugger.cpp
--- a/js/src/debugger/Debugger.cpp	Mon Jun 15 14:33:28 2020 +0000
+++ b/js/src/debugger/Debugger.cpp	Mon Jun 15 14:26:35 2020 +0000
@@ -111,7 +111,7 @@
 #include "wasm/WasmTypes.h"           // for WasmInstanceObjectVector
 
 #include "debugger/DebugAPI-inl.h"
-#include "debugger/Frame-inl.h"    // for DebuggerFrame::hasGenerator
+#include "debugger/Frame-inl.h"    // for DebuggerFrame::hasGeneratorInfo
 #include "debugger/Script-inl.h"   // for DebuggerScript::getReferent
 #include "gc/GC-inl.h"             // for ZoneCellIter
 #include "gc/Marking-inl.h"        // for MaybeForwarded
@@ -681,7 +681,7 @@
 
     if (!frames.add(p, referent, frame)) {
       frame->freeFrameIterData(cx->defaultFreeOp());
-      frame->clearGenerator(cx->runtime()->defaultFreeOp(), this);
+      frame->clearGeneratorInfo(cx->runtime()->defaultFreeOp(), this);
       ReportOutOfMemory(cx);
       return false;
     }
@@ -1204,7 +1204,7 @@
     {
       AutoRealm ar(cx, frameObj);
 
-      if (!frameObj->setGenerator(cx, genObj)) {
+      if (!frameObj->setGeneratorInfo(cx, genObj)) {
         ReportOutOfMemory(cx);
 
         // This leaves `genObj` and `frameObj` unassociated. It's OK
@@ -3833,7 +3833,7 @@
            e.popFront()) {
         DebuggerFrame* frameObj = e.front().value();
         if (IsAboutToBeFinalizedUnbarriered(&frameObj)) {
-          frameObj->clearGenerator(fop, dbg, &e);
+          frameObj->clearGeneratorInfo(fop, dbg, &e);
         }
       }
     }
@@ -4732,7 +4732,7 @@
       AbstractGeneratorObject& genObj = *e.front().key();
       DebuggerFrame& frameObj = *e.front().value();
       if (genObj.isClosed() || &genObj.callee().global() == global) {
-        frameObj.clearGenerator(fop, this, &e);
+        frameObj.clearGeneratorInfo(fop, this, &e);
       }
     }
   }
@@ -6002,7 +6002,7 @@
     if (!dbg->getFrame(cx, iter, &adoptedFrame)) {
       return false;
     }
-  } else if (frameObj->hasGenerator()) {
+  } else if (frameObj->isSuspended()) {
     Rooted<AbstractGeneratorObject*> gen(cx, &frameObj->unwrappedGenerator());
     if (!dbg->observesGlobal(&gen->global())) {
       JS_ReportErrorASCII(cx, "Debugger.Frame's global is not a debuggee");
@@ -6351,7 +6351,7 @@
     Debugger* dbg = Debugger::fromChildJSObject(frameobj);
     dbg->frames.remove(frame);
 
-    if (frameobj->hasGenerator()) {
+    if (frameobj->hasGeneratorInfo()) {
       // If this is a generator's final pop, remove its entry from
       // generatorFrames. Such an entry exists if and only if the
       // Debugger.Frame's generator has been set.
@@ -6363,7 +6363,7 @@
         MOZ_ASSERT(p);
         MOZ_ASSERT(p->value() == frameobj);
 #endif
-        frameobj->clearGenerator(fop, dbg);
+        frameobj->clearGeneratorInfo(fop, dbg);
       }
     } else {
       frameobj->maybeDecrementStepperCounter(fop, frame);